--- parser.cpp	2014-11-03 06:56:28.000000000 +0200
+++ parser.cpp	2014-11-03 06:57:35.000000000 +0200
@@ -437,7 +437,7 @@
 
 
 #if (! defined yyoverflow \
-     && (! defined __cplusplus \
+     && (! defined __cplusplus || defined GMX_YYFORCE_C_STACK_EXTENSION \
          || (defined YYLTYPE_IS_TRIVIAL && YYLTYPE_IS_TRIVIAL \
              && defined YYSTYPE_IS_TRIVIAL && YYSTYPE_IS_TRIVIAL)))
 
@@ -955,7 +955,7 @@
 static void
 yy_symbol_value_print (FILE *yyoutput, int yytype, YYSTYPE const * const yyvaluep, YYLTYPE const * const yylocationp, void *scanner)
 {
-  FILE *yyo = yyoutput;
+  FILE *yyo gmx_unused = yyoutput;
   YYUSE (yyo);
   YYUSE (yylocationp);
   YYUSE (scanner);
